utils/make_changelog.py

```diff
@@ -0,0 +1,385 @@
+        msg = info['commit']['message']
+
+        # GitHub login
+        login = info['author']['login']
+        name = None
+
+        # Firstly, try get name from github user
+        try:
+            name = users[login]['name']
+        except:
+            pass
+
+        # Then, try get name from commit
+        if not name:
+            try:
+                name = info['commit']['author']['name']
+            except:
+                pass
+
+        author = '[{}]({})'.format(name or login, info['author']['html_url'])
+        texts.append(pattern.format(commit, html_url, author, msg))
+
+    text = 'Commits which are not from any pull request:\n\n'
+    return text + '\n\n'.join(texts)
+
+
+# List of pull requests -> text description.
+def process_pull_requests(pull_requests, users, repo):
+    groups = {}
+
+    for id, item in pull_requests.items():
+        lines = list(filter(len, map(lambda x: x.strip(), item['description'].split('\n'))))
+
+        cat_pos = None
+        short_descr_pos = None
+        long_descr_pos = None
+
+        if lines:
+            for i in range(len(lines) - 1):
+                if re.match('^\**Category\s*(\(leave one\))*:*\**\s*$', lines[i]):
+                    cat_pos = i
+                if re.match('^\**\s*Short description', lines[i]):
+                    short_descr_pos = i
+                if re.match('^\**\s*Detailed description', lines[i]):
+                    long_descr_pos = i
+
+        cat = ''
+        if cat_pos:
+            # TODO: Sometimes have more than one
+            cat = lines[cat_pos + 1]
+        cat = cat.strip().lstrip('-').strip()
+
+        short_descr = ''
+        if short_descr_pos:
+            short_descr_end = long_descr_pos or len(lines)
+            short_descr = lines[short_descr_pos + 1]
+            if short_descr_pos + 2 != short_descr_end:
+                short_descr += ' ...'
+
+        # TODO: Add detailed description somewhere
+
+        pattern = u"{} [#{}]({}) ({})"
+        link = 'https://github.com/{}/pull/{}'.format(repo, id)
+        author = 'author not found'
+        if item['user'] in users:
+            # TODO get user from any commit if no user name on github
+            user = users[item['user']]
+            author = u'[{}]({})'.format(user['name'] or user['login'], user['html_url'])
+
+        if cat not in groups:
+            groups[cat] = []
+        groups[cat].append(pattern.format(short_descr, id, link, author))
+
+    texts = []
+    for group, text in groups.items():
+        items = [u'* {}'.format(pr) for pr in text]
+        texts.append(u'### {}\n{}'.format(group if group else u'[No category]', '\n'.join(items)))
+
+    return '\n\n'.join(texts)
+
+
+# Load inner state. For debug purposes.
+def load_state(state_file, base_sha, new_tag, prev_tag):
+
+    state = {}
+
+    if state_file:
+        try:
+            if os.path.exists(state_file):
+                logging.info('Reading state from {}'.format(state_file))
```

Comment:
May use %s modifier in `logging.level("foo %s", "bar")` and get rid of `format`.